book/src/design/circuit/gadgets/ecc.md

```diff
@@ -1 +1,76 @@
-# Elliptic Curve Cryptography
\ No newline at end of file
+# Elliptic Curve Cryptography
+
+## Incomplete addition
+Inputs: $P = (x_P, y_P), Q = (x_Q, y_Q)$
+Output: $A = P + Q = (x_A, y_A)$
+
+Formulae:
+- $\lambda \cdot (x_p - x_{q}) = y_p - y_{q}$
+- $x_{a} = \lambda^2 - x_{q} - x_p$
+- $y_{a} = \lambda(x_{q} - x_{a}) - y_{q}$
+
+Substituting for $\lambda$, we get the constraints:
+- $(x_{a} + x_{q} + x_p) \cdot (x_p - x_q)^2 - (y_p - y_{q})^2 = 0$
+- $(y_{a} + y_{q})(x_p - x_{q}) - (y_p - y_{q})(x_{q} - x_{a}) = 0$
+
+## Complete addition
+
+To implement complete addition inside the circuit, we need to check the following cases:
+
+$\begin{array}{rcll}
+\mathcal{O} &+& \mathcal{O} &= \mathcal{O} ✓\\
+\mathcal{O} &+& (x_q, y_q)  &= (x_q, y_q) ✓\\
+ (x_p, y_p) &+& \mathcal{O} &= (x_p, y_p) ✓\\
+   (x, y)   &+& (x, y)      &= [2] (x, y) ✓\\
+   (x, y)   &+& (x, -y)     &= \mathcal{O} ✓\\
+ (x_p, y_p) &+& (x_q, y_q)  &= (x_p, y_p) \;⸭\; (x_q, y_q), \text{if } x_p \neq x_q ✓
+\end{array}$
+
+We represent $\mathcal{O}$ as $(0, 0)$.
+
+> $0$ is not an $x$-coordinate of a valid point because we would need $y^2 = x^3 + 5$, and $5$ is not square in $\mathbb{F}_q$.
+
+$$
+\begin{aligned}
+P + Q &= R\\
+(x_p, y_p) + (x_q, y_q) &= (x_r, y_r) \\
+                \lambda &= \frac{y_q - y_p}{x_q - x_p} \\
+                    x_r &= \lambda^2 - x_p - x_q \\
+                    y_r &= \lambda(x_p - x_r) - y_p
+\end{aligned}
+$$
+
+For the doubling case, $\lambda$ has to instead be computed as $\frac{3x^2}{2y}$.
+
+Witness $\lambda, \alpha, \beta, \gamma, \delta, A, B, C, D$.
```

Comment:
If I'm reading the table correctly, we need to require that `\alpha, \beta, \gamma, \delta` are non-zero, otherwise the implied relations are only partial (e.g. `x_p = 0 \implies B` holds but `x_p != 0 \implies ¬B` does not, yet we rely on `¬B` later on). So we need additional constraints that require the prover to witness their inverses.

Reply:
@daira clarified in a pairing that it's the cross-interaction of the two B constraints that fully constraint the implications. For example, the contra-positive of the first constraint's implication `x_p == 0 ==> B` is `!B ==> x_p != 0`, which is the other half of the second constraint's implication.